File: linked_list/21.merge-two-sorted-lists/21.merge-two-sorted-lists.cpp
/*
 * @lc app=leetcode id=21 lang=cpp
 *
 * [21] Merge Two Sorted Lists
 */
struct ListNode {
     int val;
     ListNode *next;
     ListNode() : val(0), next(nullptr) {}
     ListNode(int x) : val(x), next(nullptr) {}
     ListNode(int x, ListNode *next) : val(x), next(next) {}
};

// @lc code=start
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) 
    {
        ListNode *node = nullptr, *head = nullptr, *prev = nullptr;
        while (list1 != nullptr && list2 != nullptr)
        {
            ListNode *tmp = new ListNode();
            if (list1->val < list2->val) {
                tmp->val = list1->val;                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     tmp->val = list1->val;
            } else {
                tmp->val = list2 ->val;
            }

            if (prev) {
                prev->next = tmp;
            } else {
                head = tmp;
            }

            prev = tmp;
        }

        return head;
    }
};
// @lc code=end

File: src/java/list/21.merge-two-sorted-lists.java
class P21 {
    // @lc app=leetcode id=21 lang=java
    // @lc code=start
    public class Solution {
        public ListNode mergeTwoLists(ListNode l1, ListNode l2) {
            ListNode dummy = new ListNode(0), t = dummy;
            ListNode p=l1, q=l2;
            while(p!=null || q!=null){
                if (q==null || (p!=null && p.val <= q.val)) { t.Next = p; t.next = p; p = (p!=null? (p.Next!=null?p.Next:p.next):null); }
                else { t.Next = q; t.next = q; q = (q!=null? (q.Next!=null?q.Next:q.next):null); }
                t = (t.Next!=null ? t.Next : t.next);
            }
            return dummy.Next;
        }
    }
    // @lc code=end
}
